src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs

```diff
@@ -213,51 +287,93 @@ private static void AddValueToReduceHash(object value, Hashing.Streamed.XXHash64
                     {
                         fixed (char* p = s)
                         {
-                            Hashing.Streamed.XXHash64.Process(hash64Context, (byte*)p,
-                                s.Length * sizeof(char));
+                            switch (_mode)
+                            {
+                                case Mode.SingleValue:
+                                    _singleValueHash = Hashing.XXHash64.Calculate((byte*)p, s.Length * sizeof(char));
+                                    break;
+                                case Mode.MultipleValues:
+                                    CopyToBuffer((byte*)p, s.Length * sizeof(char));
+                                    break;
+                            }
                         }
+                        
                         return;
                     }
 
                     var lcsv = value as LazyCompressedStringValue;
                     if (lcsv != null)
                     {
-                        Hashing.Streamed.XXHash64.Process(hash64Context, lcsv.Buffer, lcsv.CompressedSize);
+                        switch (_mode)
+                        {
+                            case Mode.SingleValue:
+                                _singleValueHash = Hashing.XXHash64.Calculate(lcsv.Buffer, lcsv.CompressedSize);
+                                break;
+                            case Mode.MultipleValues:
+                                CopyToBuffer(lcsv.Buffer, lcsv.CompressedSize);
+                                break;
+                        }
+
                         return;
                     }
 
                     if (value is long)
```

Comment:
Yes. I'm working on support for complex / nested objects, generally in indexes, not only here. 
In addition to the above, we need to take care of:
- storing in Lucene - probably write as JSON to allow search on them (such field can be available for full text search)
- implement `Equals` for blittable / dynamic blittable objects to properly compare i.e. GroupBy a complex object
- handling nested objects from Lucene index on querying